operator/update: Handle cluster image changes for upgrade scenario

dimitriscruz opened the pull request:
On a new redpanda image release we want the ability to upgrade the cluster. This commit focuses on handling image changes in the redpanda cluster CR by triggering a rolling update (using partitions) against the statefulset underneath the CR.

The steps are as follows: 1) check statefulset image version identical to the cluster CR; 2) if different, modify the image in the spec, in-memory; 3) perform rolling update starting from the last pod to the first; 4) in each iteration apply the update on the pod and requeue until the pod is in ready state; 5) prior to a pod update ensure that the previous pod has started listening to its Kafka API port - in a follow-up PR we will ensure the restarted process has joined its replica groups.

A new status field, Upgrading, is also introduced. This allows us to resume image updating after a re-queue as well as controller termination or crashing.

BenPope commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,253 @@
+) error {
+	f := func() (bool, error) {
+		err := r.podImageIdenticalToClusterImage(ctx, sts, image, ordinal)
+		if err != nil {
+			// Retry
+			return false, nil
+		}
+
+		return true, nil
```
This looks like:
```go
return err == nil, nil
```
It seems odd that the `err` isn't returned.

dimitriscruz replied:
Agree, but if err is returned, the backoff gets cancelled. That's how the library works.

BenPope commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,253 @@
+			return poderr
+		}
+
+		// Ideally, at this point we'd wait until all Pods are in ready state.
+		// However, if something went wrong with a previous image change and the pod
+		// is in bad state, e.g., due to wrong or erroneous image, the pod will
+		// keep restarting and require manual intervention. Instead, for now we
+		// skip the readiness requirement and "force" the image update.
```
I think it would be nice to implement this with a dummy function for now. Some considerations:

* Never take down a pod if it will result in less than a quorum of brokers to be available.
* How does/should this interact with a `PodDisruptionBudget`?
* Readiness check can be as simple as having a TCP connection on `:9092` be accepted, for now.

dimitriscruz replied:
Thanks, Ben. For now I use a Sarama consumer against the broker that just got updated (at `:9092`). I will have a separate change that will look into the group membership of the pod (and will take into account your remaining considerations).

dimitriscruz replied:
Also, I added an e2e test to start with. Another one to add is where the image is incorrect and we recover from it. I think I'll add in separate change.

BenPope commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,253 @@
+		return true, nil
+	}
+
+	return wait.ExponentialBackoff(backoff, f)
```
Should this be truncated? i.e., limited to 5 mins or something. When I've operated services in the past, I've always been tempted to kick pods manually as this can get really high, really quick.

dimitriscruz replied:
It currently has a `waitCap` of 5 minutes. It's adjustable.

BenPope reviewed (commented):
There's a fair bit of complexity here, what's the testing plan?

alenkacz commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -158,7 +167,7 @@ func (r *StatefulSetResource) Obj() (k8sclient.Object, error) {
 					InitContainers: []corev1.Container{
 						{
-							Name:		"redpanda-configurator",
+							Name:		configuratorContainerName,
```
this is actually interesting. I know it's out of scope, but how do we plan to update the configurator? I think we should note that somewhere in the document. I assume we just need to try to update the configmap and that might be it. Also rolling update the pods so that they see the new config

BenPope replied:
I think the `initContainer` as it is, should always have the same version as the application container.

It would probably be best for the operator to create the `configMap` in the `emptyDir` directly, and drop the `initContainer`. 

Almost all of the configuration can be moved to rpk redpanda start args, except node-id and seeds - an example is in the [helm-charts](https://github.com/vectorizedio/helm-charts/pull/25/commits/e38731b257984d4228c49eb2af2785048d6298e7#diff-f74d1889126f524c4b7e7e7836a376c58b0ca5aabe3720e03732b0e25d6ee380R71)

Given that Redpanda configuration is having a massive overhaul, it might not be worth the effort for now.

BenPope replied:
> oh we use configmap to also bundle the shell script - I was mainly talking about that. Sorry I should have made it more clear https://github.com/vectorizedio/redpanda/blob/dev/src/go/k8s/pkg/resources/configmap.go#L39

Ahh, I see. In my previous company we developed [Wave](https://github.com/wave-k8s/wave) to automagically kick pods based on changes to config - I think it supports statefulSets although I never used it for them, but the gist of it is the same.

dimitriscruz replied:
Init container is updated along the main container.

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,253 @@
+
+		r.logger.Info("Waiting for pod to complete updating", "partition:", ordinal)
+
+		if err := r.waitUntilPodImageIsVerified(ctx, sts, newImage, ordinal); err != nil {
```
one concern I have about this approach that the whole update process blocks the reconciliation worker which is problematic and usually not ideal approach because potentially other changes can stall depending on size of the cluster and amount of workers you have.

Have you thought about making this asynchronous? Could we perhaps watch on pod updates and proceed based on that?

dimitriscruz replied:
The wait was removed. Resolving this.

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,253 @@
+		// Give time to redpanda pod to rejoin cluster.
+		// In a revision, an API call to redpanda is made to help us decide when to continue.
+		r.logger.Info("Sleeping for 1 minute before updating the next pod, or completion")
+		time.Sleep(time.Minute)
```
oh wow, I would really like to make this asynchronous somehow...

alenkacz replied:
we should absolutely at least have some timeout here but then when we hit that timeout, we let the sts in kind of weird state so we need to make sure that the upgrade phase can be continued after that...

dimitriscruz replied:
Yeah, this will be removed (I thought I did already), and be replaced with logic towards the start of the upgrade. I'll update the commit these days.

dimitriscruz replied:
@alenkacz I updated the commit. There's no sleep() anymore. I also added a call to the broker's Kafka API that I will update in another commit to check the previously updated pod has joined all its groups. This goes beyond the "READY" state that k8s provides. (I don't add it as a condition to the pod's status because we don't always want it to be a condition for a pod to be ready).

alenkacz reviewed (changes requested):
I have a huge concern mainly about the upgrade process blocking the worker for pretty big amount of time. Let's have a discussion around that.

This is lacking tests as well.

dimitriscruz commented:
> I have a huge concern mainly about the upgrade process blocking the worker for pretty big amount of time. Let's have a discussion around that.

@alenkacz let me know if the updated approach looks better. I introduced an `Upgrading` state so the controller isn't blocked and it can also survive controller crashes. 

I removed the sleep that imitated the Kafka API call but haven't replaced it yet, you may ignore that part for now. I'll figure it out separately.

dotnwat reviewed (commented):
What is the purpose of tracking an `Upgrading` state explicitly?

alenkacz reviewed (changes requested):
Left couple more comments, I think we're heading in the right direction :)

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
```
I think this logic definitely deserves also unit test coverage

RafalKorepta replied:
As I wrote comment in `updateStsImage` the logic of rolling update should be anchor solely on the following attributes of STS.Status:

- UpdatedReplicas
- CurrentReplicas
- ReadyReplicas
- Replicas

It should find the next ordinal. Reschedule the reconciliation loop. See if it can continue with next ordinal.

dimitriscruz replied:
Rafal, because we are doing partition rolling update, I'm not sure we can use UpdatedReplicas. When we provide partition i, only nodes from i to N-1 will be updated. So the UpdatedReplicas might be misleading. I'd need to check the behavior to be certain. At the same time, since we're contemplating switching to non-statefulset, it'd probably be better to not overly depend on such properties and deal with pods directly during upgrade. This way, if we switch away from StatefulSet the upgrade path will need require fewer changes.

dimitriscruz replied:
Alena, thanks for the suggestion. I will be adding testing, including e2e, once we settle on the implementation. I hope the next few days.

alenkacz replied:
I actually reopened this PR to give a similar suggestion as Rafal did, to use the sts status properties :) I think it would be easier and easier to test.

But I get that point with moving out of sts (possibly) ... so I am fine with that. But yeah, actually with this we're replicating a lot of the logic that sts controller is already doing for us. One more reason to not rely on that in the future :-)

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -46,6 +46,9 @@ type ClusterStatus struct {
 	Nodes	[]string	`json:"nodes,omitempty"`
+	// Indicates cluster is upgrading
+	// +optional
+	Upgrading	bool	`json:"upgrading"`
```
actually I am fine with keeping this and evolving it into state in following PR or whatever is ready first, we're in alpha API stage anyway

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -90,6 +95,10 @@ func (r *StatefulSetResource) Ensure(ctx context.Context) error {
 		}
 	}
 
+	if err := r.updateStsImage(ctx, &sts); err != nil {
```
You don't need `&sts` as receiver holds LastObservedState.

dimitriscruz replied:
I use `sts` because otherwise I'd just add `sts := r.LastObservedState` in each function, so I preferred to make it obvious that we operate on the sts by passing it around.

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+
+const requeueDuration = time.Second * 10
+
+func (r *StatefulSetResource) updateStsImage(
```
I see this function differently. To embrace returning early and don't nesting unnecessarily the pseudo code should look like:
``` go
func (r *StatefulSetResource) updateStsImage(ctx context.Context, sts *appsv1.StatefulSet,) error {
    if images in sts and cluster CR are the same and ready replicas are the same as cluster.spec.replicas {
        return
    }
    
    set the upgrade status in the cluster CR
    
    modifyPodImage
    
    lunchOnePodUpdate using `StatefulSet's partition number`

    return to reschedule
}
```

We can argue that setting upgrade status and modify pod image could be done once. Never the less this should be the core logic behind this function.

dimitriscruz replied:
The replicas condition wouldn't work. I explained in an earlier comment-reply you made. Beyond the if-statement, what you describe is the logic currently implemented.

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	if container.Image != newImage {
+		r.logger.Info("Container image not updated to cluster image", "pod", pod.Name,
```
This log statement can be bubble up the stack to the main reconciliation loop using returned error.

dimitriscruz replied:
Keep in mind that not all errors returned result in an error reaching the reconciliation loop. They might be handled on the way there. Such errors would not be logged. This is an example of such case.

RafalKorepta replied:
Agree with you

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+	return nil
+}
+
+func (r *StatefulSetResource) podImageIdenticalToClusterImage(
```
You don't need to check all of this, because STS will report readyReplicas in its status.

dimitriscruz replied:
I gave my explanation in a similar comment you gave previously.

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+		time.Sleep(time.Minute)
+
+		// After Update(), refresh the statefulset for the next iteration (or other use).
+		err := r.Get(ctx, r.Key(), sts)
```
Maybe use `r.LastObservedState` instead of sts?

dimitriscruz replied:
I use `sts` because otherwise I'd just add `sts := r.LastObservedState` in each function, so I preferred to make it obvious that we operate on the sts by passing it around.

RafalKorepta reviewed (changes requested):
Looks promising. In order not to obscure review I'm adding few review request change here:

- Add more context to the returned errors. Please follow golang blog post https://blog.golang.org/go1.13-errors. TLDR `fmt.Errorf("wail doing something: %w", err)`
- Revisit log level in some of the function as it start to get to verbose for me at least
- Don't log before returning an error. Add context to the error and let the main reconciliation loop log the information.
- Create tests. It would be the best if they are E2E.

dimitriscruz commented on src/go/k8s/controllers/redpanda/cluster_controller.go:
```diff
@@ -82,6 +82,13 @@ func (r *ClusterReconciler) Reconcile(
 	for _, res := range toApply {
 		err := res.Ensure(ctx)
+
+		var e *resources.NeedToReconcileError
```
I don't mind introducing a "finished" bool. The disadvantage is we'd be adding a bool to the return of a chain of functions. Is the current approach going against a convention, or is there a concrete disadvantage you see?

dimitriscruz replied:
You're talking about `log.Info`? :)

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+
+	if upgrading || rpContainer.Image != newImage {
+		podSpec := &sts.Spec.Template.Spec
+		if err := r.modifyPodImage(podSpec, newImage); err != nil {
```
Yeah

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+	}
+
+	if upgrading {
+		r.logger.Info("Continuing cluster image update", "cluster image", newImage)
```
Sure, I'll make this one a 4.

dimitriscruz replied:
Actually, we'd need to consider all levels, so in this PR I'd rather not set any levels. The operator has no levels at the moment.

RafalKorepta replied:
Initially it had. Probably I didn't pay much attention to it, so it get lost. Let's leave it for next PR.

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+
+	newImage := fmt.Sprintf("%s:%s", r.pandaCluster.Spec.Image, r.pandaCluster.Spec.Version)
+	if rpContainer.Image != newImage {
+		r.logger.Info("Starting cluster image update", "cluster image", newImage, "container image", rpContainer.Image)
```
Yeah, check out something similar: https://github.com/microsoft/azure-databricks-operator/blob/265a0a8546293a5d52cdb8a98bac507ce0750cae/controllers/djob_controller.go#L83

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,277 @@
+	if !reflect.DeepEqual(upgrading, r.pandaCluster.Status.Upgrading) {
+		r.pandaCluster.Status.Upgrading = upgrading
+		if err := r.Status().Update(ctx, r.pandaCluster); err != nil {
+			r.logger.Error(err, "Failed to update RedpandaClusterStatus")
```
We have similar Update code in other parts, so let's change all at once.

RafalKorepta replied:
Ok agree. Next PR then.

dimitriscruz commented:
> What is the purpose of tracking an `Upgrading` state explicitly?

@dotnwat Advantages I see: 1) user can see that the cluster is upgrading without inferring it through the status or image version of each pod. This improves user experience. 2) we don't have to loop over the pods to establish whether we are upgrading. Since almost always we won't be upgrading that would be somehow wasteful, although the cost is probably unimportant.

These advantages do not make introducing an `Upgrading` state necessary for correctness. (Are they good enough?). A disadvantage would be that we introduce a bug in tracking the state and then the upgrade doesn't continue as expected..

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,309 @@
+	conf.ClientID = "operator"
+	conf.Admin.Timeout = time.Second
+
+	consumer, err := sarama.NewConsumer(addresses, conf)
```
Doesn't like it, but It connects to the cluster.

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,309 @@
+			Msg:	fmt.Sprintf("wait for pod (ordinal: %d) to restart", ordinal)}
+	}
+
+	// Ensure 0th Pod is ready for I/O before completing the upgrade.
+	if err := r.ensureRedpandaGroupsReady(sts, replicas, 0); err != nil {
+		return &NeedToReconcileError{RequeueAfter: requeueDuration,
+			Msg:	fmt.Sprintf("redpanda on pod (ordinal: %d) not ready", 0)}
+	}
```
Because after we update each pod in the loop, we requeue. If the pod becomes ready in the meantime we'd have no chance to call `ensureRedpandaGroupsReady` for it.

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,309 @@
+
+	consumer, err := sarama.NewConsumer(addresses, conf)
+	if err != nil {
+		logger.Error(err, "Error while creating consumer")
```
We have a couple of instances in the cluster controller
https://github.com/vectorizedio/redpanda/blob/aa0a2371f10f22a90f1421cc14031d4a75703799/src/go/k8s/controllers/redpanda/cluster_controller.go#L111

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,309 @@
+		r.logger.Info("Continuing cluster image update", "cluster image", newImage)
+	}
+
+	if upgrading || rpContainer.Image != newImage {
```
sure, I can change that.

RafalKorepta commented on src/go/k8s/controllers/redpanda/cluster_controller.go:
```diff
@@ -82,6 +82,13 @@ func (r *ClusterReconciler) Reconcile(
+
+		var e *resources.NeedToReconcileError
+		if errors.As(err, &e) {
+			log.Info(e.Error())
```
Ok after discussion let's leave the `log.info`

RafalKorepta reviewed (approved):
LGTM

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,310 @@
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
```
so what if the pod is updated but it did not join the group and we retriggered reconcile on line 113, that's incorrect, right?

dimitriscruz replied:
Please help me out understand this. Here is a list of invariants:

- Each iteration "group-verifies" the _previously_ updated pod
- Each iteration updates 0 or 1 pod
- A pod is updated only if the _previous_ pod is "group-ready".

Note that `r.ensureRedpandaGroupsReady(sts, replicas, ordinal+1)` checks `ordinal+1` not `ordinal`.

alenkacz replied:
Oh you're right, then I guess we're fine. I missed that part, perhaps we could make it a bit more obvious but that could be refactoring after. Thanks for the explanation

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,310 @@
+
+// Ensures the Redpanda pod has rejoined its groups after restarting,
+// i.e., is ready for I/O.
+func (r *StatefulSetResource) ensureRedpandaGroupsReady(
```
this looks to me like something we can make in the future into readiness/health check, do you agree? Or am I misunderstanding this?

dimitriscruz replied:
Yeah, definitely. Either within the operator or as an endpoint in Redpanda. It's in my list. :)

alenkacz replied:
Should we maybe start opening github issues for these things rather than everyone keeping their lists? I don't want all this to be forgotten :) I can do that as well if you don't want to.

dimitriscruz replied:
I can add it under Project when we use that.

alenkacz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,310 @@
+	return queryRedpandaForTopicMembers(addresses, r.logger)
+}
+
+// To be implemented: currently only connects to Kafka API.
```
nit: remove this

alenkacz replied:
maybe add some context to this? I guess the assumption is that if the consumer can connect, the groups are ready?

dimitriscruz replied:
Unfortunately it's more complex to verify that all groups area read. Here we're saying "at least we can connect and retrieve topics information" - the retrieval happens when `NewConsumer` is called. I will be adding the required logic in this function, or might introduce a /health endpoint in redpanda.

BenPope replied:
I'd have to check the logic, but I suspect sarama will try to connect to all brokers that have joined the cluster, so it might not do quite what you expect, but you probably get away with it when things are going well, at least.

dimitriscruz replied:
@BenPope I'm not providing those brokers to sarama though.. I'm providing the one to test.

@alenkacz: I updated the comment.

dimitriscruz replied:
Note that I'll be revisiting this function and "health" overall.

BenPope replied:
It doesn't care - they're seeds, it will query for all the brokers and then connect to them (maybe). Same reason you can't use rpk to do this - it uses sarama, anyway.

alenkacz replied:
btw. this is also problematic when we introduce TLS because then controller will have to have certs to do this call right?

alenkacz reviewed (changes requested):
I still have couple of requests:
- mostly for documentation
- I think there's one place in which the reconcile loop is not correct (that it might happen that the groups are not rejoined yet)

Once we resolve this I think it's good to go. The rest is just a discussion and follow-ups we can do after this merges.

But nice work :clap: I am so happy we ended up with fully async loop in the end, thank you!

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -90,6 +97,10 @@ func (r *StatefulSetResource) Ensure(ctx context.Context) error {
 		}
 	}
 
+	if err := r.updateStsImage(ctx, &sts); err != nil {
```
I'd think that any change restarting the cluster containers should go through a partitioned rolling update (to avoid availability issues). We might need to abstract the upgrade logic here (and pass functions) to cater for other kinds of updates.

alenkacz replied:
yeah, since that's a bug in my code let's merge this and adjust the resource update to do that. Opened #656 thanks

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset_update.go:
```diff
@@ -0,0 +1,310 @@
+
+const requeueDuration = time.Second * 10
+
+func (r *StatefulSetResource) updateStsImage(
```
Makes sense, added.

alenkacz commented:
I think my only remaining concern is that with the kafka api call directly from the controller we'll need to have certs when TLS is introduced as that complicates things, otherwise 👍
